Rfc 1 draft wip (#73)
tango-controls/rfc: 1/README.md

shortname: 1/TANGO
name: Tango Control System
status: raw
editor: Lorenzo Pivetta (lorenzo.pivetta@elettra.eu)
contributor: Andy Götz (andy.gotz@esrf.fr)
---

> This document describes Tango, a control system framework. This RFC define the specification of a control system implementing the philosophy of Tango.

Reviewer: ```suggestion
This document describes Tango, a control system framework. This RFC defines the specification of a control system implementing the philosophy of Tango.
```

See also: Y/OtherTemplate

## Preamble

> Copyright (c) 2019 European Synchrotron (ESRF).

Reviewer 2: Shouldn't this read "Copyright (c) 2019 Tango Controls" ?

Author: Right!

Reviewer 3: Damm you discovered my secret plan!

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](http://tools.ietf.org/html/rfc2119).

## Tango Specification

> The TANGO control system is a toolkit for building distributed control systems which are device oriented. 

Reviewer 2: Likely, this going to be the very first section one would read when coming to Tango RFC... suggest to explain a little bit the concept "device oriented", e.g. "... which are device oriented, modelling the hardware equipment or the software component into the device abstraction". I think this can be improved how it is stated, though.

### Goals

> Tango aims to be extremely flexible and can be used to interface any kind of hardware or software system which can be accessed via one of the software languages supported by Tango (C++, Python, Java and JavaScript).

Reviewer 2: s/accessed/interfaced/

The goal of Tango is to provide a way of representing all physical and software systems which need to be controlled by a control system as Tango Devices.

Devices are software components which have Commands (methods), Attributes (data) and Properties (configuration) and are accessible on the network or locally.

> A Tango control system consists of a collection of Tango Devices, an (optional) Database and a set of applications for configuring and controlling the Devices via well defined Application Programmers Interfaces (APIs). Tango defines and maintains official APIs for C++, Python, Java, JavaScript and REST.

Reviewer 2: s/Application Programmers Interface/Application Programming Interface/
Also, not sure to state "(optional) Database" here... I'm in the idea Tango requires the Database, and working without is just a workaround for limited setups and use cases...

> The Tango Database serves the purpose of defining and declaring all Devices in the control system and their configuration parameters. The Database also stores the network address of each Device and serves as Naming Service for clients who need to establish a connection to any or all Devices.

Reviewer 2: I'd remove "stores the network addresses" as it is likely to me misinterpreted here... rephrase as "The Database also serves as Naming Service for clients who need to establish a connection to any or all Devices".

The software used in large facilities can be compared with the software for stock markets – huge amount of data must be displayed on the monitor in a real time and being processed and being saved in databases for post processing.

> The Tango toolkit was mainly developed for research facilities needs, but the idea and concept (philosophy) behind was to create a framework for control systems in general. The concept is based on the idea that the concept of an object or component is ideally suited to designing and implementing control systems. TANGO was based on this concept and can be described as an object oriented control system. A number of concepts were defined as being necessary to implement a TANGO control system (ref "TANGO - AN OBJECT ORIENTED CONTROL SYSTEM BASED ON CORBA" by Chaize et. al., ICALEPCS 1999). These are:

Reviewer 2: I'd rephrase this as:
"The Tango toolkit was mainly developed for research facilities needs, but the idea and concept (philosophy) behind was to create a framework for control systems in general. It is based on the idea that distributed objects are really suited to design and implement control systems. TANGO was based on this concept, and can be described as an Object Oriented distributed control system. A number of requirements were defined in the original design (ref... TANGO - An Object ...):

1. ...
2. ...

1. Device network object with the minimum of functionality necessary for a distributed control object
2. Use CORBA communication protocol while hiding details
3. Support following communication paradigms - synchronous, asynchronous and event driven
4. Support the implementation of new devices by sub-classing Device and by composing new types of Devices from existing types of Devices
5. Support modern programming patterns like multi-threading
6. A naming database which stores configuration information

> 7. A generic set of Data Types

Reviewer 2: Suggest to rephrase "A generic, but limited, set of Data Types"

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](http://tools.ietf.org/html/rfc2119).

## Tango Specification
The TANGO control system is a toolkit for building distributed control systems which are device oriented. 

### Goals

> Tango aims to be extremely flexible and can be used to interface any kind of hardware or software system which can be accessed via one of the software languages supported by Tango (C++, Python, Java and JavaScript).

Reviewer 4: I would somehow mark that the intention is to support other languages, too.

> The goal of Tango is to provide a way of representing all physical and software systems which need to be controlled by a control system as Tango Devices.

Reviewer 4: ```suggestion
The goal of Tango is to provide a way of representing all physical and software systems which need to be controlled by a control system as Tango Devices (see [RFC 2/Device](/2)).
```

> Devices are software components which have Commands (methods), Attributes (data) and Properties (configuration) and are accessible on the network or locally.

Reviewer 4: ```suggestion
Devices are software components which have [Commands](/3) (methods), [Attributes](/4) (data) and [Properties](/5) (configuration) and are accessible on the network or locally.
```

Reviewer 4: Should Pipes be listed here?

A Tango control system consists of a collection of Tango Devices, a Tango Database and a set of applications for configuring and controlling the Devices via well defined Application Programmers Interfaces (APIs). Tango defines and maintains official APIs for C++, Python, Java, JavaScript and REST.

> The Tango Database serves the purpose of defining and declaring all Devices in the control system and their configuration parameters. The Database also stores Naming Service for clients who need to establish a connection to any or all Devices.

Reviewer 4: ```suggestion
The Tango Database serves the purpose of defining and declaring all Devices in the control system and their configuration parameters. The Database also stores Naming Service for clients who need to establish a connection to any or all Devices (see [RFC 6/Database](/6)).
```

Additionally, it aims to:

> * Provide an Open Source software toolkit for building distirbuted object oriented control systems. It should be freely available for all applications including industry. Tango Controls is sufficiently complete to control any small or large installation. Provide the necessary libraries for 3 programming languages which allow hardware to be interfaced and modelled in a device object model. Provide the tools for configuring, managing, monitoring and archiving large numbers of control points implemented as device attributes. Provide naming, browsing and hierarchies of devices and groups of devices.

Reviewer: ```suggestion
* Provide an Open Source software toolkit for building distributed object oriented control systems. It should be freely available for all applications including industry. Tango Controls is sufficiently complete to control any small or large installation. Provide the necessary libraries for 3 programming languages which allow hardware to be interfaced and modeled in a device object model. Provide the tools for configuring, managing, monitoring and archiving large numbers of control points implemented as device attributes. Provide naming, browsing and hierarchies of devices and groups of devices.
```

Reviewer: I suggested 2 typos fixes (distributed and modeled)... Not so easy to see with github interface it seems.
I think you can put 1 sentence per line, it will be easier to suggest changes. If you don't end up your lines with 2 spaces, it will result in the same presentation.
